### PR TITLE
refactor: histogram

### PR DESCRIPTION
**Pull request by chux0519**
Open the PR.
Still have problem to be solved.
See the FIXME part in code

### CODE REVIEW

**Comment by chux0519** [comment]
I just implemented generic version of some types in `histogram.rs`.
It will break one test case, because of the f64 to i64 casting.
I'm not sure how to fix this. is it OK to change the `duration_to_seconds` function to like `duration_to_millis`. Accurate to milliseconds and it should pass all tests.

@breeswish

**Comment by chux0519** [comment]
related issue: 
1. https://github.com/tikv/rust-prometheus/issues/254
2. https://github.com/tikv/tikv/issues/5704

File: src/histogram.rs [hunk 1]
```diff
@@ -397,21 +408,29 @@ impl HistogramTimer {
         let v = duration_to_seconds(self.start.elapsed());
```
Comment (chux0519) [inline src/histogram.rs]: Accurate to milliseconds should fix this, is it OK to do so? @breeswish

File: src/histogram.rs [hunk 2]
```diff
@@ -397,21 +408,29 @@ impl HistogramTimer {
...
         self.observed = true;
         if record {
-            self.histogram.observe(v);
+            // FIXME: 精度丢失
```
Comment (siddontang) [inline src/histogram.rs]: don't use Chinese comments

File: src/histogram.rs [hunk 3]
```diff
@@ -394,24 +405,30 @@ impl HistogramTimer {
...
 }
 
-/// A [`Metric`](::core::Metric) counts individual observations from an event or sample stream in
+/// The underlying implementation for [`Histogram`](::Histogram) and [`IntHistogram`](::IntHistogram).
+#[derive(Debug)]
+pub struct GenericHistogram<S: Atomic, C: Atomic> {
+    core: Arc<HistogramCore<S, C>>,
+}
+
+/// /// A [`Metric`](::core::Metric) counts individual observations from an event or sample stream in
```
Comment (chux0519) [inline src/histogram.rs]: duplicate `///` mark

File: src/histogram.rs (hunk 2 above)
Comment (chux0519) [reply src/histogram.rs]: sure, fixed in last commit

**Comment by breeswish** [comment]
@chux0519 Thanks for the PR! Looks like your PR makes a breaking change that previously `Histogram` reports durations in seconds but now it reports in milliseconds. You may consider adding a new interface that records millisecond precision.

**Comment by chux0519** [comment]
> @chux0519 Thanks for the PR! Looks like your PR makes a breaking change that previously `Histogram` reports durations in seconds but now it reports in milliseconds. You may consider adding a new interface that records millisecond precision.

you are right, I will definitely fix this

**Comment by mxinden** [comment]
@chux0519 are you planning to continue the work on this pull request? If not, I would close here.

**Comment by chux0519** [comment]
> @chux0519 are you planning to continue the work on this pull request? If not, I would close here.

Sorry, I may not be able to continue this PR, it can be closed

**Pull request closed without merging**